app/src/main/java/com/google/android/apps/watchme/StreamerActivity.java

```diff
@@ -160,10 +168,72 @@ private void startStreaming() {
         wakeLock.acquire();
 
         if (!streamerService.isStreaming()) {
-            streamerService.startStreaming(rtmpUrl);
+
+            String cameraPermission = Manifest.permission.CAMERA;
+            String microphonePermission = Manifest.permission.RECORD_AUDIO;
+            int hasCamPermission = checkSelfPermission(cameraPermission);
+            int hasMicPermission = checkSelfPermission(microphonePermission);
+            List<String> permissions = new ArrayList<String>();
+            if (hasCamPermission != PackageManager.PERMISSION_GRANTED) {
+                permissions.add(cameraPermission);
+                if (ActivityCompat.shouldShowRequestPermissionRationale(this,
+                        Manifest.permission.CAMERA)) {
+                    // Provide rationale in Snackbar with button to request permission
+                    Snackbar.make(preview, R.string.permission_camera_rationale,
+                            Snackbar.LENGTH_INDEFINITE).show();
+                }
+            }
+            if (hasMicPermission != PackageManager.PERMISSION_GRANTED) {
```

Comment:
We should account for the permission refused case as well - e.g. if they have been asked but the user said 'no', tell them to go turn it on in the settings dialog.